### PR TITLE
GEOPY-787

### PR DESCRIPTION
**Pull request by domfournier**
**GEOPY-787 - Base joint inversion driver (cross-gradient regularization)**

### CODE REVIEW

File: geoapps/inversion/joint/joint_cross_gradient/driver.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+    def __init__(self, params: JointCrossGradientParams):
+        self._wires = None
+        self._directives = None
+
+        super().__init__(params)
+
+        with fetch_active_workspace(self.workspace, mode="r+"):
+            self.initialize()
+        #
```
Comment (benk-mira) [inline geoapps/inversion/joint/joint_cross_gradient/driver.py:42]:
```suggestion
        
```

File: geoapps/inversion/joint/joint_cross_gradient/driver.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+                "length_scale_z",
+                "s_norm",
+                "x_norm",
+                "y_norm",
+                "z_norm",
+                "gradient_type",
+            ]:
+                if getattr(self.params, param) is None:
+                    for reg in reg_block:
+                        print(param, getattr(driver.params, param))
```
Comment (benk-mira) [inline geoapps/inversion/joint/joint_cross_gradient/driver.py:147]: cleanup

File: geoapps/inversion/joint/joint_cross_gradient/driver.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+        return self._wires
+
+    @property
+    def mapping(self):
+        """Create a list of all mappings augmented with projection to global problem"""
+        if self._mapping is None:
+            mappings = {}
+            start = 0
+            n_values = int(np.sum(self.models.active_cells))
```
Comment (benk-mira) [inline geoapps/inversion/joint/joint_cross_gradient/driver.py:199]:
```suggestion
            n_values = self.models.n_active
```

File: geoapps/inversion/joint/joint_cross_gradient/driver.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+            self._mapping = mappings
+
+        return self._mapping.values()
+
+    @property
+    def n_values(self):
+        """Number of values in the model"""
+        if self._n_values is None:
+            n_values = int(np.sum(self.models.active_cells))
```
Comment (benk-mira) [inline geoapps/inversion/joint/joint_cross_gradient/driver.py:215]:
```suggestion
            n_values = self.models.n_active
```

File: geoapps/inversion/joint/joint_cross_gradient/params.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+from copy import deepcopy
+
+from geoapps.inversion.joint.params import BaseJointParams
+
+from .constants import default_ui_json, inversion_defaults, validations
+
+
+class JointCrossGradientParams(BaseJointParams):
+    """
+    Parameter class for gravity->density inversion.
```
Comment (benk-mira) [inline geoapps/inversion/joint/joint_cross_gradient/params.py:19]:
```suggestion
    Parameter class for cross-gradient joint inversion.
```

File: geoapps/inversion/joint/joint_cross_gradient/params.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    @property
+    def physical_property(self):
+        """Physical property to invert."""
+        return self._physical_property
+
+    @physical_property.setter
+    def physical_property(self, val: list[str]):
+        # unique_properties = list(set(val))
+        # if len(unique_properties) == len(self.):
+        #     raise ValueError(
```
Comment (benk-mira) [inline geoapps/inversion/joint/joint_cross_gradient/params.py:50]: cleanup

File: geoapps/inversion/driver.py [hunk 7]
```diff
@@ -505,6 +513,7 @@ def get_path(self, filepath: str | Path) -> str:
 
 
 if __name__ == "__main__":
-    file = str(Path(sys.argv[1]).resolve())
+    # file = str(Path(sys.argv[1]).resolve())
```
Comment (benk-mira) [inline geoapps/inversion/driver.py:516]:
```suggestion
```

**Review by benk-mira — CHANGES_REQUESTED** [review]
mostly just cleanups.  Well done man!

File: geoapps/inversion/driver.py [hunk 8]
```diff
@@ -189,6 +196,14 @@ def models(self):
 
         return self._models
 
+    @property
+    def n_values(self):
+        """Number of values in the model"""
+        if self._n_values is None:
+            self._n_values = int(np.sum(self.models.active_cells))
```
[the remark this answers is not in the document]
Comment (domfournier) [reply geoapps/inversion/driver.py:203]: Cool yep. It is `n_active` right now in `models`, but that works.

File: geoapps/inversion/joint/joint_cross_gradient/driver.py (hunk 3 above)
Comment (domfournier) [reply geoapps/inversion/joint/joint_cross_gradient/driver.py:199]: Same, yep.

File: geoapps/inversion/driver.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (benk-mira) [reply geoapps/inversion/driver.py]: whoops, that's what i meant!

**Review by benk-mira — APPROVED** [review]
Love it!

**Pull request merged**